### ISSUE TITLE
Combined injections not entirely highlighted

### ISSUE THREAD

**Issue by ribru17** [neovim/neovim#31777]
### Problem

The treesitter highlighter fails to consider all text of a combined injection, even when it is visible on screen. Sometimes it will draw it correctly but the highlights go back to being incorrect on the next redraw(?)

### Steps to reproduce

minimal.lua:

```lua
vim.treesitter.query.set(
  'lua',
  'injections',
  [[((comment_content) @injection.content (#set! injection.self) (#set! injection.combined))]]
)

vim.api.nvim_buf_set_lines(0, 0, -1, false, {
  '-- print([[some',
  '-- text',
  '-- here]])',
})
vim.bo.ft = 'lua'
```

```sh
nvim --clean -u minimal.lua
```

Observe:
![Image](https://github.com/user-attachments/assets/d3de9396-f1c5-424b-9cd4-8ad22da6b023)

Weird redrawing:
https://github.com/user-attachments/assets/8e845f91-29fe-46ae-a3bd-617e2d8387cd

### Expected behavior

In the above example, `here]]` should also have a string highlight, and the final closing parenthesis should have a bracket highlight. Instead they are not parsed as injections.

### Nvim version (nvim -v)

NVIM v0.11.0-dev-1422+g35247b00a4-dirty Build type: Debug LuaJIT 2.1.1734355927

### Vim (not Nvim) behaves the same?

NA

### Operating system/version

Linux 6.12.5 @ NixOS

### Terminal name/version

kitty 0.37.0

### $TERM environment variable

xterm-kitty

### Installation

Build from source

**Comment by vanaigr** [comment]
Possibly related: https://github.com/neovim/neovim/pull/30869#issuecomment-2430570060
(Very first highlight in a tree may be drawn out of order)

**Comment by ribru17** [comment]
Thanks, that is interesting. Very odd. A fix right now is to lower the priority of the root tree marks by one (`highlight_spec` also still passes) but it seems hacky

**Comment by vanaigr** [comment]
I recently tried to fix a different issue (https://github.com/neovim/neovim/pull/31778#issue-2762390570) andy first attempt was to precalculate all possoble priority values child trees can take, and then offset everything so that the priorities don't overlap.

I giess doing that would also fix this issue.

**Comment by ribru17** [comment]
That makes sense. It does seem overly complex though... also I once tried tackling the scoped injection issue and in my opinion it would make more sense to handle this using quantified query patterns; do you think so? As in nodes of `(node)+ @injection.content` would all be scoped as one. Also because adding a new capture like `@injection.root` is partially limited by upstream https://tree-sitter.github.io/tree-sitter/3-syntax-highlighting.html#language-injection

**Comment by vanaigr** [comment]
Pulled relevant parts into its own branch:
https://github.com/vanaigr/neovim/tree/dc0cdb01b370baa3602bc323c23b1e54a9906cb9

will try to may a PR later, but it would need further discussion. Some highlights rely on being obscured by the parent tree.

**Comment by ribru17** [comment]
Yeah the cases you say where highlights rely on parent tree obscuring make me nervous... thinking about it again would just adding 1 to child tree priorities be that bad? It at least solves this issue and the change is much simpler, some like https://github.com/neovim/neovim/commit/2b515d9e26b2789e49d357796904f98aefb17d69
